text/0076-rawkv-batch-export.md

```diff
@@ -0,0 +1,138 @@
+# RFC: RawKV Batch Export
+
+- RFC PR: https://github.com/tikv/rfcs/pull/76
+- Tracking Issue: https://github.com/tikv/tikv/issues/11007
+- [Proposal: RawKV SST Export (Chinese)](https://docs.google.com/document/d/1IpraWthiFBWKHeDVX3x7aQ6X7pqzknHlfCQpyaMDBWQ/edit)
+
+## Summary
+
+Let RawKV support Batch Export.
+
+## Motivation
+
+More and more users want to use RawKV as online store in [the feature store](https://www.featurestore.org/).
+
+But two important functions are missing:
+
+- data import, which is done in [Online Bulk Load](https://github.com/tikv/tikv/issues/10563).
+- data export, which we want to provide in this RFC.
+
+`RawKV Batch Export` should meet the following requirements:
+
+1. the export process should be as fast as possible,
+2. the impact to the online serving (P99 latency) should be under control,
+3. the export speed can be improved in a linear way by adding more tikv instances,
+4. the export framework should be able to process Big Data (500GB ~ 5TB) in a short period (from minutes to several hours).
+
+## Detailed design
+
+The solution proposed in this RFC is as follows:
+
+1. Using [RawKV backup](https://docs.pingcap.com/zh/tidb/stable/use-br-command-line-tool#raw-kv-%E5%A4%87%E4%BB%BD%E5%AE%9E%E9%AA%8C%E6%80%A7%E5%8A%9F%E8%83%BD) provide by BR, TiKV can backup SST files to S3. But HDFS is not supported yet, which is describe in [Support upload SST to HDFS](#support-upload-sst-to-hdfs).
+2. After uploading SST files to S3 / HDFS, we will use Spark to decode the SST files to Key-Value pairs with the help of [the RocksDB SstFileReader API](https://github.com/facebook/rocksdb/blob/6.23.fb/java/src/main/java/org/rocksdb/SstFileReader.java). And then the Key-Value pairs will be convert to user firendly file formats like parquet, which can be controlled by user code. After that the parquet file will finally upload to S3 / HDFS.
```

Comment:
Proposal 1: [HDFS/S3] <-- Parquet file -- [Spark Job] <-- [SST on HDFS/S3]-- [TiKV Cluster] <-- SST Files -- [Disk]

Proposal 2: [HDFS/S3] <-- Parquet file -- [Spark Job] <-- key-value stream -- [TiKV Cluster] <-- SST Files -- [Disk]

These two proposals both work.

Reasons for Proposal 1:
- `Backup SST files to S3` is already provided by BR. We do not need to develop another code in Spark which is a little bit complicated.
- RawKV Batch Export will benefit if the BR backup logic is optimized (to decrease the impact on online serving).
- Do not need to maintain two code logics that do the same things.

Reasons for Proposal 2:
- It's simple. One Spark Job can finish the export task.

Reply:
I'd like to compare the two proposals in the following aspects:

Functionality:
- export speed on the same cluster
- performance impact on the QPS and P99 latency in the same export speed

Stability:
- the bulk export speed and performance impact can be steady during bulk exporting process

Usability:
- reduce the client side work as much as possible

Cost:
- the network cost
- the storage cost
- the CPU cost

Reply:
I prefer the second proposal for the following reasons:

1. despite the export speed and impact on the online service, the usability is better and the cost is less than the first proposal.
2. bulk export is symmetric with bulk load, the second implementation can be easier to understand because it's also symmetric with the implementation of bulk load.